Cache Auto-configuration

snicoll opened the pull request:
Here's the initial support for cache auto-configuration aimed at Spring Boot 1.3 (#2633). This is a joint effort with @eddumelendez

The following cache libraries are supported:
* Generic (via the presence of `org.springframework.cache.Cache` instances  in the context)
* EhCache
* Hazelcast
* JSR-107 compliant provider
* Redis
* Guava
* Simple (i.e. concurrent Map)
* NoOp 

By default, the environment is checked to determine which cache libraries should be used. In the case of ehcache and hazelcast, we check if the library is present of course but also if either a default configuration file is present (`/ehcache.xml`, `/hazelcast.xml`) or if a custom location has been set (via the common `spring.cache.config` property).

The `spring.cache.mode` allows to specify the caching library to use in case of conflict. It should be noted that only `noop` should be set explicitly. Maybe `redis` should be in that situation as well (right now the presence of a `RedisTemplate` enable a cache on redis, not sure this is appropriate).

`spring.cache.cache-names` defines the comma-separated list of caches to create on startup. 

We need to add several samples to showcase how this works (and validate that the conditions work has they should with different classpath content).

jskjons commented:
would be cool if this supported infinispan too

snicoll commented:
Yup. I think the main goal here is to validate the architecture and design decisions. We can consider adding more later, if appropriate.

philwebb commented:
@snicoll I made a start on merging this but there's a few things I'd like to discuss first. My working branch is here https://github.com/philwebb/spring-boot/tree/gh-2633-all

1) I think we might need to support `CachingConfigurer` detection with the conditions. If you specify a `CacheManager` or `CacheResolver` using a `CachingConfigurer` then they won't be `@Beans` and the existing conditions won't kick in.

2) Consider an enum for `CacheProperties.type` (previously `mode`). I think it might be good to use an enum so that we get auto-completion hints in the IDE. Not quite as easy to extend but I think that's probably OK.

3) It would be nice to rework `CacheValidationConfiguration` somehow (although I don't know how). It's unusual to have a configuration that always fails. Perhaps we could do the test with a `BeanFactoryPostProcessor`?

4) `GuavaCacheConfiguration.createCacheManager()` The if/then/else logic in this method is a little confusing, it might be worth extracting the first set to a private method. Not sure what it should be called.

I'll pick this up again tomorrow and we can rebase into a sane commit history before we merge.

Cheers!

snicoll commented:
Thanks for the review, I added a few comments on your commits. 

Good catch for `CachingConfigurer`, I should know better! I had a look (your new condition should be named `ConditionalOnMissingCacheConfiguration` IMO) and I tried to implement the condition. The easiest would be to inject the `CacheAspectSupport` and check if `getCacheResolver()` returns `true`. I tried that but then I ended up with early initialization problems. A much less nicer option would be to check if a `CachingConfigurer` bean exists and then look at the outcome of `cacheManager()` and `cacheResolver` but that's duplicating a bit of the logic. I could consider adding a boolean flag on that class to indicate if a "cache configuration is set" (no idea how to call that getter).

Mode was an enum initially and then I change it to a simple String. Here's why:

1. We have a precedent with `shell.auth` (I am not saying it's a good one).
2. We can expect IDE to catch up and offer auto-completion for such values (when we finally work on the domain thing in the meta-data). Right now the value is auto-completed in upper case which is less nicer and a reason why I don't like using them too much. 

I am ok if we go with an enum though.

`CacheValidationConfiguration` is a weird beast. I added that at the very end because without it, the check in `@EnableCaching` barfs at you without giving any insight of what could have been gone wrong. I like this idea of having a "last ressort check" that throws a dedicated exception. I don't like how it's implemented though.

The if/else logic in `GuavaCacheConfiguration.createCacheManager()` can be refactored in a method named `setCacheBuilder` because this is really what it is about: `CacheBuilder`, `CacheBuilderSpec` and the specification string are 3 ways of initializing a `CacheBuilder`

snicoll commented:
One more thing, I don't think that add that new condition in each `XYZCacheConfiguration` is a good idea. When `CacheAutoConfiguration` is evaluated, it should back-off if:

1. A `CacheManager` bean is set
2. A `CacheManager` or ` CacheResolver` has been set in `CacheAspectSupport` in one way or the other (typically via `CacheConfigurer` but you can also take the crazy use case of someone creating the interceptor manually. Spring Data JPA does such a thing for the `TransactionInterceptor`).

Once those conditions are validated to be false, you only need to check if a `CacheManager` bean is set in each nested configuration, repeating those checks above is useless IMO.

philwebb commented:
Looking at `CachingConfigurer` the Javadoc states that implemented methods should also add `@Bean` annotations. I think that this means we can continue to just check for missing beans and ignore the fact that `AbstractCachingConfiguration` can setup resolvers/managers even if they aren't beans.

snicoll commented:
Yeah, the bad news about that is that I have no idea why this was specified that way. Could it be to prevent a double call to the method? having the `@Configuration` proxy caching the method calls?

philwebb commented:
@snicoll The comments predate your work on this, perhaps it's technically not required any more. I think it's probably a fine compromise to say you need to define a `CacheManager` or `CacheResolver` as `@Beans` if you don't want auto-configuration.

snicoll commented:
I had more thoughts today on that and I still believe that the only viable solution is to look at `CacheAspectSupport#getCacheResolver`. If we can't do that, then that compromise is fine I guess.

philwebb commented:
Merged with tweaks in commits 151220f41d1108c9b84f98c6d8eddc710a0d5863 and 48959f25533bf59c2511165d4c801b8716da7467
